### ISSUE TITLE
Covid-19 Lab Certification / City of Oberhausen/Germany

### ISSUE THREAD

**Issue by rrz-dev** [hzi-braunschweig/SORMAS-Project#5318]
### Situation Description
The city of Oberhausen wants to use SORMAS for tracking and planning of Covid-19 cases and laboratory tests.

The city of Oberhausen Fire Department is planning Covid-19 testing at utility facilities. A lab order must be generated for the sample to be tested. This should be generated with Sormas and the necessary information should be stored in the task so that it can be easily processed.

We would appreciate the opportunity to discuss these requirements with a project manager (preferably in German). First attempts to contact the SORMAS team at the start of development via EMail were unfortunately rejected due to lack of time.

### Feature Description
We extended the task form and data structure with all necessary fields for lab certification generation. We added a new action to the form to create a PDF form, that is printable by the user. We extended the task list with some new quick filters and mass editing actions.

As the features and data is only usefull in germany, all new fields are limited to germany.

### Additional Information
The city of Oberhausen has commissioned Aleri Solutions GmbH with the implementation. The feature has already been developed and the resulting code is to be submitted to the open source repository. This feature can be used by various other cities in germany, as they have similar requirements.

**Comment by MateStrysewske** [comment]
@JaquM @kwa20 Can you suggest how to move forward with this issue? We'd basically need a functional confirmation before we could merge this into the development branch. I guess it would probably be necessary for @rrz-dev to very specifically list the development specifications here (i.e. which exact fields are added where, what is added to the UI where, etc.) for you to evaluate that, right?

@rrz-dev Concerning the technical approach - there are quite a lot of things I've seen in the PR where I don't agree with the approach. For example, since this seems to be an issue that is very specific to Germany, adding so many additional fields e.g. to the task is not really an option. We might need an individual entity for that instead. I'll go more into detail once the functional part of this issue has been discussed.

**Comment by kwa20** [comment]
@MateStrysewske @rrz-dev Yes, a detailed specification would help a lot to evaluate what has been done.

**Comment by Candice-Louw** [comment]
@rrz-dev - do you have an update on delivering this document, please?  You are welcome to contact me directly [candice.louw(at)helmholtz-hzi.de] if it is easier.

**Comment by rrz-dev** [comment]
I'm looking forward to provide the information until tomorrow.

**Comment by Candice-Louw** [comment]
@rrz-dev - any update yet, please?

**Comment by rrz-dev** [comment]
These are the changes we added to SORMAS to meet the requirements of the city of Oberhausen:

The requirement is, that the employees of city of Oberhausen needs to generate and print "lab certificates" when planning "sample collection tasks". This print has to be send with the sample to the laboratory to identify it.

For this requirement we changed the Task detail view. We added all necessary edit fields for the "lab certificate" but only, when TaskType.SAMPLE_COLLECTION is selected.

**Changed**: _Task detail view_
- added editing for all new TaskDto fields for TaskType.SAMPLE_COLLECTION

This fields need to be saved to the database for generating and printing the "lab certificates" at a later time. For this we extended the TaskDto / Entity with all the needed fields.

**Changed**: _TaskDto / Entity_
- new Fields to generate labCertificates, needed when planning a "sample collection task"
- field list:
-- labCertificateGuid
-- payerNumber
-- doctorNumber
-- operatingFacilityNumber
-- labNumber
-- testV
-- selfPaying
-- specialAgreement
-- firstTest
-- nextTest
-- contactPerson
-- coronaApp
-- outbreak
-- outbreakPrevention
-- workingInFacility
-- livingInFacility
-- medicalFacility
- communityFacility
-- careFacility
-- otherFacility
-- agreedToGdpr
-- specialAgreementCode
-- healthDepartment (Reference)

The "lab certificate" is issued by a health department of a region. For this we created a new FacilityType "Health Department" to be able to add this to the SORMAS settings, make it selectable when creating a "lab certificate" and changed the FacilityEdit form to meet this requirement.

**Changed**: added new FacilityType "Health Department"

**Changed**: _FacilityEdit form_
- added all new fields of FacilityDto

Some more information was needed for the new health department facility, so we extended the FacilityDto / Entity to  store this information in the database.

**Changed**: _FacilityDto / Entity_
- field list:
-- Department
-- Sector
-- Dr_name
-- house_no
-- tel_no
-- fax_no

For working with a bigger number of tasks we extended the task list with new actions and filters.

**Changed**: _Task list / grid_
- new actions (mass editing): markAsDone and printLabornachweise (lab certificates)
- new filters: "due date today", "due date tomorrow"

To enable the new functions only to a specific user group we added the UserRole "Fire department".

**Changed**: UserRole
- added Role FIRE_DEPARTMENT

The generation of "lab certificates" is done by a newly added, external java dependency: "de.aleri.lab-certificate-generator". In this library all the pdf generation of "lab certificates" is encapsulated regarding to the german health department regulations.

**Comment by Marko-ilmkreis** [comment]
@rrz-dev I think you referencing to that form: https://www.kbv.de/media/sp/Ansichtsexemplar_MusterOEGD.pdf

We are generating these forms every morning before our test teams are going to the test center or starting the tour to take samples at the home address of the cases/contacts.

For us it would not be needed to print the form at the desk of the inventigator for cases/contacts as the location of the test-team is in different building and we are having a lot of immidiatley tests every day. E.g. if a kid in school or kindergarden was testet by rapid test, we are making a PCR the same day.

Maybe this issue can solved by having the detailed export (#3792, #3106) together with serial printing within word?
